ctapipe/image/time_next_neighbor_cleaning.py

```diff
@@ -0,0 +1,742 @@
+
+    def dump_individual_pixel_rates(self, file):
+        """
+        Dump the IPR graph to file using json.
+
+        Parameters
+        ----------
+        file: string
+
+        """
+        with open(file, "w") as f:
+            json.dump(self.IPR_dict, f, sort_keys=True, indent=4)
+
+    def fill_individual_pixel_rate(self, geo, image, trace_length, charge_steps):
+        """
+        Count the number of pixels in images about a given charge threshold
+        and fill dictionary with individual pixel noise rate (IPR).
+
+        Parameters
+        ----------
+        geo: ctapipe.instrument.CameraGeometry
+        image: numpy.ndarray
+        trace_length: astropy.units.Quantity
+            total length of the trace, e.g. number of samples times sample_time
+        charge_steps: numpy.ndarray
+            array with the charge steps used to calculate the IPR graphs.
+        """
+
+        self.IPR_dict = self.IPR_dict if self.IPR_dict else {}
+
+        if geo.camera_name not in self.IPR_dict:
+            self.IPR_dict[geo.camera_name] = {
+                "counter": 0,
+                "charge": charge_steps,
+                "npix": np.zeros(charge_steps.shape[0]),
+                "rate": np.zeros(charge_steps.shape[0]),
+                "rate_err": np.zeros(charge_steps.shape[0]),
+            }
+
+        # count number of pixels
+        self.IPR_dict[geo.camera_name]["counter"] += np.sum(image >= 0)
+        for thbin, charge in enumerate(self.IPR_dict[geo.camera_name]["charge"]):
+            self.IPR_dict[geo.camera_name]["npix"][thbin] += np.sum(image > charge)
+
+        # convert count of number of pixels above threshold into rate
+        sample_to_hz = (
+            1 / (trace_length * self.IPR_dict[geo.camera_name]["counter"])
+        ).to("Hz")
+        self.IPR_dict[geo.camera_name]["rate"] = (
+            self.IPR_dict[geo.camera_name]["npix"] * sample_to_hz
+        )
+        self.IPR_dict[geo.camera_name]["rate_err"] = (
+            np.sqrt(self.IPR_dict[geo.camera_name]["npix"]) * sample_to_hz
+        )
+
+    def calculate_ipr_from_source(self, source, calibrator, charge_steps):
+        """
+        Fill the IPR graph from a file or a list of files. The files should
+        contain pure night sky background calibration data. Therefore, the
+        calibration of the events in the source are performed. The calibrator
+        for this case should be equal to the one in the analysis.
+
+        Parameters
+        ----------
+        source:
+        calibrator:
+        charge_steps: numpy.ndarray
+            array storing the steps in charge used to create the IPR graph
+
+        """
+        for event in source:
+            calibrator.calibrate(event)
+            for tel_id in event.r0.tels_with_data:
+                geometry = event.inst.subarray.tel[tel_id].camera
+                image = event.dl1.tel[tel_id].image[0]
+
+                n_samples = event.r0.tel[tel_id].num_samples
```

Comment:
I am not sure it is correct to use the waveform num_samples and the sampling frequency here to estimate the rate.

Also: shouldn't the rate be dimensionless? I.e. just be the percentage of events that survive a given charge threshold?

@kpfrang

Reply:
If I remember correctly, `sum_time` should really be the total duration (with units) of the trace. If this is correct, the calculation of `sum_time = n_samples * sample_time` seems correct to me.

This calculation is to generate fill the `IPR_dict`  which is equivalent to the graphs that are shown in Fig. 1 of the reference: 
https://galprop.stanford.edu/elibrary/icrc/2013/papers/icrc2013-0655.pdf

Reply:
I don't understand why a time is needed at all. The goal is to have a "rate" of X percent pedestal events surviving, right?

So why do we need a "real" rate with some timeunit?

Reply:
You can have a look at equations 1 and 2 of the proceeding. I think it gets clear from this.

Seems that I tried to convert the fake probability (which is the input parameter) into the accidental rate R_acc from these formulas in the line below. I don't remember how this is done in EventDisplay though. Probably worth checking this.
https://github.com/cta-observatory/ctapipe/blob/8d614a9a77c1d7b11523b22c9b67764e7c733bb6/ctapipe/image/time_next_neighbor_cleaning.py#L648

Reply:
Here is a inverse cumulative plot of the pedestal pixel values, pretty much reproduces the figure (without any assumptions about "time"):

![pixel_values](https://user-images.githubusercontent.com/5488440/159904441-5adb9795-aa6c-4fae-8e59-df4b4d3ed078.png)

Reply:
@kpfrang Do you remember which extractor you used? If it was the `NeighborPeakWindowSum` I think the times of neighboring pixels are correlated and thus the time range rejection might not work as expected, since that assumes uncorrelated times. Maybe that explains that more pedestal events than your target  rate you survived?

Reply:
Not quite sure, but I believe I used the `NeighborPeakWindowSum`.